rfd/headless-login.md

```diff
@@ -0,0 +1,367 @@
+
+Note: Ideally, we would ensure that the only client that can *receive* certificates is the requesting client, but ensuring the PKCE flow above is enough to defend against attacks since the certificates are useless without the client's private key. If we want to remove the possibility of an attack attempting to intercept the certificates, we can use an HTTP websocket or gRPC stream endpoint to encapsulate the headless authentication request into a single API request. This would require some additional implementation details to be flushed out, so it is not in this initial draft design.
+
+### Headless authentication overview
+
+The headless authentication flow is shown below:
+
+```mermaid
+sequenceDiagram
+    participant Local Machine
+    participant Remote Machine
+    participant Teleport Proxy
+    participant Teleport Auth
+
+    Note over Remote Machine: tsh --headless ssh user@node01
+    Remote Machine->>+Teleport Proxy: initiate headless login
+    Teleport Proxy->>Teleport Auth: save request with new uuid
+    Note over Teleport Auth: Request: {id, user, publicKey, ...}
+    Teleport Auth->>Teleport Proxy: request ID
+    Teleport Proxy->>-Remote Machine: request ID
+    Note over Remote Machine: print command + URL
+
+    Remote Machine->>+Teleport Proxy: request signed user certs<br/>for request ID
+    Note over Local Machine: user copies command / URL<br/>to local terminal / browser
+    Local Machine->>+Teleport Proxy: initiate headless request approval
+    Teleport Proxy->>Local Machine: MFA challenge
+    Note over Local Machine: user taps YubiKey
+    Local Machine->>Teleport Proxy: signed MFA challenge response
+    Teleport Proxy-->>-Local Machine: success/failure
+
+    Teleport Proxy->>Teleport Auth: Request signed user certificates
+    Teleport Auth->>Teleport Proxy: Issue signed user certificates<br/>for the client's public key 
+    Teleport Proxy->>-Remote Machine: User certificates
+    Note over Remote Machine: Connect to user@node01
+```
+
+This flow can be broken down into three parts: headless login initiation, local authentication, and certificate retrieval.
+
+#### Headless login initiation
+
+First, the client initiates a headless login. In this stage, the client provides request details including the Teleport username, client public key, and the command being requested (e.g. `tsh ssh user@host`). The request will begin in the pending state, to be later approved/denied by the user locally.
+
+```go
+type HeadlessLoginRequest struct {
+    // User is a teleport username
+    User string `json:"user"`
+    // PubKey is a public key user wishes to sign in
+    PubKey []byte `json:"pub_key"`
+    // Command is the client command being requested
+    Command string
```

Comment:
How can we make sure that the command sent here is the command that is being executed?

Reply:
Since this is a client provided field, we can't make sure it is called from the server side. The command is only meant to be used to provide additional details for the user to see and verify before approving the request. The request ID is the primary, non-client-tamperable thing that should be user verified though.